### PR TITLE
Codebreacker v.1.0.0.0

### PR DESCRIPTION
**Pull request by Wizard2007**

### CODE REVIEW

File: lib/codebreaker.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+require 'codebreaker/version'
+
+require 'codebreaker/game'
+require 'codebreaker/console_service'
+require 'yaml'
+module Codebreaker
+  module Yaml_io
```
Comment (yamnich) [inline lib/codebreaker.rb:7]: names of classes and modules should be capitalized camel case

File: lib/codebreaker/console_service.rb [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    def get_user_input
+      user_input = String.new(gets.chomp())
+      (user_input != '') ? process_user_input(user_input) : process_user_input(:empty_input)
+    end
+    def process_user_input(user_input)
+      if (!valid_answer?(user_input) && @restaet_was_called)
+        send_user_input(:restart_game)
+        return
+      end
+      if @current_game.secret_code_valid?(user_input) then
```
Comment (yamnich) [inline lib/codebreaker/console_service.rb:33]: no then

File: lib/codebreaker.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    def appned(file_name)
+      yaml_string = self.to_yaml
+      yaml_string += File.read file_name if File.exist?(file_name)
+      File.write(file_name, yaml_string)
+    end
+  end
+  class Player
```
Comment (yamnich) [inline lib/codebreaker.rb:14]: one empty line between methods , classes, modules

File: lib/codebreaker/console_service.rb [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+      puts 'ok we will start new game'
+    end
+    def no
+      @start_new_game = false
+      @restaet_was_called = false
+      puts 'ok we will close your session'
+    end
+    def play_game
+      begin
+        if @current_game.user_vin == true
```
Comment (yamnich) [inline lib/codebreaker/console_service.rb:89]: 1. typo - user_vin 
2. as usual it is better to do smth like user_win?  method

File: lib/codebreaker/console_service.rb [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+          your_vin
+          restart_game
+          return
+        end
```
Comment (yamnich) [inline lib/codebreaker/console_service.rb:93]: each part can be moved to method

File: lib/codebreaker/game.rb [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    
+    def generate_secret_code 
+      digits = [1,2,3,4,5,6]
+      @secret_code = digits.sample(code_size).join('')
+    end
+    def game_over?
+      @step_count > @max_step_count
+    end
+    def secret_code_valid?(code = '')
+      return false if code == nil
```
Comment (yamnich) [inline lib/codebreaker/game.rb:29]: code.nil?

File: lib/codebreaker/game.rb [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    end
+
+    def check_code(code)
+      result = ''
+      @step_count += 1
+      @user_vin = false
+      if @secret_code == code
+        @user_vin = true
+        return result = '++++'
+      end
```
Comment (yamnich) [inline lib/codebreaker/game.rb:42]: can be a method

File: lib/codebreaker/game.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      @step_count > @max_step_count
+    end
+    def secret_code_valid?(code = '')
+      return false if code == nil
+      l_secret_code = @secret_code
+      l_secret_code = code if code != ''
+      l_secret_code.match( /[1-6]+/) ? true : false
+    end
+
+    def check_code(code)
```
Comment (yamnich) [inline lib/codebreaker/game.rb:35]: method is too long

File: lib/codebreaker/game.rb [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+      result = ''
+      @step_count += 1
+      @user_vin = false
+      if @secret_code == code
+        @user_vin = true
+        return result = '++++'
+      end
+      l_code = String.new(@secret_code)
```
Comment (yamnich) [inline lib/codebreaker/game.rb:43]: what is meaning l_code?

File: lib/codebreaker/game.rb [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+            l_include_index = 0
+          end 
+        end
+        if l_include_index == 0
+          l_code[index] = '*'
+        else 
+          l_code[l_include_index] = '*'
+        end
+        code[index] = '*'
+      end
```
Comment (yamnich) [inline lib/codebreaker/game.rb:62]: 1. too many if/else, too complex
2. why we need marking as * here?

File: lib/codebreaker/game.rb [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        return result = '++++'
+      end
+      l_code = String.new(@secret_code)
+      code.chars.each_with_index do |element, index|
+        l_include_index = 0
+        if element == l_code[index]
+          result += '+'
+        else
+          l_include_index = l_code.index(element)
+          if l_include_index
```
Comment (yamnich) [inline lib/codebreaker/game.rb:50]: short enough to use one string statement (? :)

File: lib/codebreaker/game.rb [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        code[index] = '*'
+      end
+      result
+    end
+   
+    def play_game
+    end
+
+    def get_hint_digit
+      puts @secret_code
```
Comment (yamnich) [inline lib/codebreaker/game.rb:70]: why we need puts here?

File: spec/console_service_spec.rb [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+        expect(subject).to receive(:send).with(:show_validation_result,user_input)
+        subject.process_user_input(user_input)
+      end
+      it 'not validated by game and call send_user_input(user_input)' do
+        subject.start
+        user_input = 'vvv'
+        expect(subject).to receive(:send_user_input).with(user_input)
+        subject.process_user_input(user_input)
+      end
+      [:test].each do |name|
```
Comment (yamnich) [inline spec/console_service_spec.rb:95]: do we need each here?

File: spec/console_service_spec.rb [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+        it "should test call process_user_input and call send_user_input :unknown_command #{name}" do
+          subject.start
+          expect(subject).to receive(:send).with(:unknown_command, name.to_s)
+          subject.process_user_input(name)
+        end
+      end
+
+      [['1111','1111',"validation result : ++++\n"], ['1234','4321',"validation result : ----\n"] , ['1245','1254',"validation result : ++--\n"]].each do  |l_secret_code, l_code, l_out_put|
```
Comment (yamnich) [inline spec/console_service_spec.rb:103]: this is not enough examples

File: lib/codebreaker/game.rb (hunk 9 above)
Comment (Wizard2007) [reply lib/codebreaker/game.rb:43]: local variable for storing copy of @secret_code

**Pull request closed without merging**